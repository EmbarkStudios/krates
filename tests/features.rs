mod util;

/// Ensures weak dependencies are properly pruned if not explicitly pulled in
/// https://github.com/EmbarkStudios/krates/issues/41
#[test]
fn prunes_multiple_weak_features() {
    let mut cmd = krates::Cmd::new();
    cmd.manifest_path("tests/features/Cargo.toml")
        .features(
            ["blocking", "json", "multipart", "stream"]
                .into_iter()
                .map(|f| f.to_owned()),
        )
        .no_default_features();

    let mut builder = krates::Builder::new();
    builder.include_targets([("x86_64-unknown-linux-gnu", vec![])]);
    let md: krates::Krates = builder.build(cmd, krates::NoneFilter).unwrap();

    for name in md.krates().map(|k| &k.name) {
        println!("{name}");
    }

    // All of the specified features have a weak dependency on reqwest, so it
    // shouldn't be present in the graph
    assert_eq!(0, md.krates_by_name("reqwest").count());
    assert_eq!(0, md.krates_by_name("reqest").count());
}

/// While the zlib features brings in git2, the openssl dependency for both
/// git2 and git2-sys is optional and weak, and since we've not explicitly
/// enabled a feature to bring it in, it should not be present in the graph,
/// even though `cargo metadata` will list it in the graph
#[test]
fn prunes_mixed_dependencies() {
    let mut cmd = krates::Cmd::new();
    cmd.manifest_path("tests/features/Cargo.toml")
        .features(["zlib".to_owned()]);

    let mut builder = krates::Builder::new();
    builder.include_targets([("x86_64-unknown-linux-gnu", vec![])]);
    let md: krates::Krates = builder.build(cmd, krates::NoneFilter).unwrap();

    assert_eq!(0, md.krates_by_name("openssl-sys").count());
    // cmake is brought in via the zlib-ng-compat feature. gross.
    assert_eq!(1, md.krates_by_name("cmake").count());
}

macro_rules! assert_features {
    ($graph:expr, $name:expr, $features:expr) => {
        let (_, krate) = $graph.krates_by_name($name).next().unwrap();

        assert_eq!($graph.get_enabled_features(&krate.id).unwrap(), $features);
    };
}

/// Ensures we can enable crate features even when that crate has been renamed
/// as well as features being precise
#[test]
fn handles_features_for_renamed_crates() {
    let mut cmd = krates::Cmd::new();
    cmd.manifest_path("tests/features/Cargo.toml")
        .no_default_features()
        .features(["midi".to_owned()]);

    let mut builder = krates::Builder::new();
    builder.include_targets([("aarch64-apple-darwin", vec![])]);
    let md: krates::Krates = builder.build(cmd, krates::NoneFilter).unwrap();

    assert_eq!(1, md.krates_by_name("coreaudio-rs").count());

    // Ensure only the single feature is enabled, not all of them
    assert_features!(md, "coreaudio-sys", ["core_midi"]);
}

/// Ensures that explicitly toggling on an optional dependency works
#[test]
fn handles_explicit_weak_features() {
    {
        let mut cmd = krates::Cmd::new();
        cmd.manifest_path("tests/features/Cargo.toml")
            .no_default_features()
            .features(["reqest".to_owned()]);

        let mut builder = krates::Builder::new();
        builder.include_targets([("x86_64-unknown-linux-musl", vec![])]);
        let md: util::Graph = builder.build(cmd, krates::NoneFilter).unwrap();

        let dg = krates::petgraph::dot::Dot::new(md.graph()).to_string();

        std::fs::write("ohno.dot", dg).unwrap();

        // assert_features!(
        //     md,
        //     "reqwest",
        //     [
        //         "__rustls",
        //         "__tls",
        //         "async-compression",
        //         "brotli",
        //         "cookie_crate",
        //         "cookie_store",
        //         "cookies",
        //         "hyper-rustls",
        //         "proc-macro-hack",
        //         "rustls",
        //         "rustls-pemfile",
        //         "rustls-tls",
        //         "rustls-tls-webpki-roots",
        //         "tokio-rustls",
        //         "tokio-util",
        //         "webpki-roots"
        //     ]
        // );
    }

    // {
    //     let mut cmd = krates::Cmd::new();
    //     cmd.manifest_path("tests/features/Cargo.toml")
    //         .no_default_features()
    //         .features(["reqest".to_owned()]);

    //     let mut builder = krates::Builder::new();
    //     builder.include_targets([("x86_64-unknown-linux-musl", vec![])]);
    //     builder.ignore_kind(krates::DepKind::Normal, krates::Scope::All);
    //     let md: krates::Krates = builder.build(cmd, krates::NoneFilter).unwrap();

    //     assert_features!(
    //         md,
    //         "reqwest",
    //         ["cookie_crate", "cookie_store", "cookies", "proc-macro-hack"]
    //     );
    // }

    // let mut cmd = krates::Cmd::new();
    // cmd.manifest_path("tests/features/Cargo.toml")
    //     .features(["reqest".to_owned()]);

    // let mut builder = krates::Builder::new();
    // builder.include_targets([("x86_64-unknown-linux-musl", vec![])]);
    // let md: krates::Krates = builder.build(cmd, krates::NoneFilter).unwrap();

    // assert_features!(md, "reqwest", ["json", "cookies"]);
}
