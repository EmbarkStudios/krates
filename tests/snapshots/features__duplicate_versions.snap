---
source: tests/features.rs
expression: dotgraph
snapshot_kind: text
---
digraph {
    0 [ label = "crate bitflags 1.3.2" ]
    1 [ label = "crate defmt 0.3.8" ]
    2 [ label = "crate defmt-macros 0.3.9" ]
    3 [ label = "crate defmt-parser 0.3.4" ]
    4 [ label = "crate embedded-hal 0.2.7" ]
    5 [ label = "crate embedded-hal 1.0.0" ]
    6 [ label = "crate feature-bug-2 0.1.0 path+file:///krates/tests/feature-bug-2" ]
    7 [ label = "crate nb 0.1.3" ]
    8 [ label = "crate nb 1.1.0" ]
    9 [ label = "crate proc-macro-error 1.0.4" ]
    10 [ label = "crate proc-macro-error-attr 1.0.4" ]
    11 [ label = "crate proc-macro2 1.0.89" ]
    12 [ label = "crate quote 1.0.37" ]
    13 [ label = "crate syn 1.0.109" ]
    14 [ label = "crate syn 2.0.87" ]
    15 [ label = "crate thiserror 1.0.69" ]
    16 [ label = "crate thiserror-impl 1.0.69" ]
    17 [ label = "crate unicode-ident 1.0.13" ]
    18 [ label = "crate version_check 0.9.5" ]
    19 [ label = "crate void 1.0.2" ]
    20 [ label = "feature default" ]
    21 [ label = "feature unstable" ]
    22 [ label = "feature default" ]
    23 [ label = "feature default" ]
    24 [ label = "feature default" ]
    25 [ label = "feature full" ]
    26 [ label = "feature extra-traits" ]
    27 [ label = "feature default" ]
    28 [ label = "feature unstable" ]
    29 [ label = "feature unproven" ]
    30 [ label = "feature defmt-03" ]
    31 [ label = "feature proc-macro" ]
    32 [ label = "feature proc-macro" ]
    33 [ label = "feature defmt-03" ]
    34 [ label = "feature defmt" ]
    35 [ label = "feature defmt-03" ]
    36 [ label = "feature default" ]
    37 [ label = "feature syn-error" ]
    38 [ label = "feature syn" ]
    39 [ label = "feature proc-macro" ]
    40 [ label = "feature printing" ]
    41 [ label = "feature parsing" ]
    42 [ label = "feature derive" ]
    43 [ label = "feature clone-impls" ]
    1 -> 20 [ label = "" ]
    1 -> 2 [ label = "" ]
    2 -> 21 [ label = "" ]
    2 -> 22 [ label = "" ]
    2 -> 23 [ label = "" ]
    2 -> 24 [ label = "" ]
    2 -> 25 [ label = "" ]
    2 -> 26 [ label = "" ]
    2 -> 27 [ label = "" ]
    3 -> 15 [ label = "" ]
    4 -> 7 [ label = "" ]
    4 -> 28 [ label = "" ]
    4 -> 19 [ label = "" ]
    5 -> 1 [ label = "" ]
    6 -> 1 [ label = "" ]
    6 -> 29 [ label = "" ]
    6 -> 5 [ label = "" ]
    6 -> 30 [ label = "" ]
    7 -> 8 [ label = "" ]
    9 -> 10 [ label = "" ]
    9 -> 23 [ label = "" ]
    9 -> 24 [ label = "" ]
    9 -> 13 [ label = "" ]
    9 -> 18 [ label = "(build)" ]
    10 -> 23 [ label = "" ]
    10 -> 24 [ label = "" ]
    10 -> 18 [ label = "(build)" ]
    11 -> 17 [ label = "" ]
    12 -> 11 [ label = "" ]
    12 -> 31 [ label = "" ]
    13 -> 11 [ label = "" ]
    13 -> 17 [ label = "" ]
    14 -> 11 [ label = "" ]
    14 -> 31 [ label = "" ]
    14 -> 12 [ label = "" ]
    14 -> 32 [ label = "" ]
    14 -> 17 [ label = "" ]
    15 -> 16 [ label = "" ]
    16 -> 23 [ label = "" ]
    16 -> 24 [ label = "" ]
    16 -> 27 [ label = "" ]
    20 -> 0 [ label = "" ]
    21 -> 3 [ label = "" ]
    29 -> 4 [ label = "" ]
    29 -> 28 [ label = "" ]
    30 -> 5 [ label = "" ]
    30 -> 33 [ label = "" ]
    34 -> 6 [ label = "" ]
    34 -> 1 [ label = "" ]
    34 -> 35 [ label = "" ]
    36 -> 6 [ label = "" ]
    28 -> 7 [ label = "" ]
    37 -> 9 [ label = "" ]
    37 -> 38 [ label = "" ]
    38 -> 9 [ label = "" ]
    38 -> 13 [ label = "" ]
    22 -> 9 [ label = "" ]
    22 -> 37 [ label = "" ]
    31 -> 11 [ label = "" ]
    23 -> 11 [ label = "" ]
    23 -> 31 [ label = "" ]
    32 -> 12 [ label = "" ]
    32 -> 31 [ label = "" ]
    24 -> 12 [ label = "" ]
    24 -> 32 [ label = "" ]
    39 -> 14 [ label = "" ]
    39 -> 31 [ label = "" ]
    40 -> 14 [ label = "" ]
    40 -> 12 [ label = "" ]
    41 -> 14 [ label = "" ]
    25 -> 14 [ label = "" ]
    26 -> 14 [ label = "" ]
    42 -> 14 [ label = "" ]
    27 -> 14 [ label = "" ]
    27 -> 42 [ label = "" ]
    27 -> 41 [ label = "" ]
    27 -> 40 [ label = "" ]
    27 -> 43 [ label = "" ]
    27 -> 39 [ label = "" ]
    43 -> 14 [ label = "" ]
}
